**Loading and Checking data**

In [2]:
import pandas as pd

# Load dataset
df = pd.read_csv("/content/bank.csv")

# Preview first rows
print(df.head())

# Number of rows and columns
print("\nShape:", df.shape)

# Column names
print("\nColumns:")
print(df.columns)

# Data types + missing values
print("\nInfo:")
print(df.info())

# Summary statistics
print("\nSummary statistics:")
print(df.describe())


   age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes  

Shape: (11162, 17)

Columns:
Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'h

**Missing and Unknown Values**

In [3]:
#handling missing data
print (df.isnull().sum())

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
deposit      0
dtype: int64


In [5]:
print((df == "unknown").sum())

age             0
job            70
marital         0
education     497
default         0
balance         0
housing         0
loan            0
contact      2346
day             0
month           0
duration        0
campaign        0
pdays           0
previous        0
poutcome     8326
deposit         0
dtype: int64


In [6]:
#convert unknown into missing
import numpy as np
print (df.replace("unknown", np.nan, inplace=True))

None


In [7]:
print (df.isnull().sum())

age             0
job            70
marital         0
education     497
default         0
balance         0
housing         0
loan            0
contact      2346
day             0
month           0
duration        0
campaign        0
pdays           0
previous        0
poutcome     8326
deposit         0
dtype: int64


In [9]:
# fixing missing values
df["job"] = df["job"].fillna("Not Specified")
df["education"] = df["education"].fillna("Not Specified")
df["contact"] = df["contact"].fillna("Not Specified")
df["poutcome"] = df["poutcome"].fillna("No Previous Campaign")


In [10]:
print (df.isnull().sum())

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
deposit      0
dtype: int64


**Duplicates**

In [11]:
#checking for duplicates
print (df.duplicated().sum())

0


**Checking the Pdays**

In [12]:
# pdays count
df["pdays"].value_counts()

,count
pdays,
-1,8324
92,106
182,89
91,84
181,81
...,...
717,1
159,1
118,1


In [15]:
#convert -1 as missing
import numpy as np
df["pdays"] = df["pdays"].replace(-1, np.nan)

In [16]:
df["pdays"].isnull().sum()

np.int64(8324)

In [17]:
# creating contact history
df["previously_contacted"] = df["pdays"].notnull()

In [18]:
df["previously_contacted"].value_counts()

,count
previously_contacted,
False,8324
True,2838


**Exploratory Data Analysis**

In [19]:
# checking deposit (yes / no)
df["deposit"].value_counts()

,count
deposit,
no,5873
yes,5289


In [20]:
df["deposit"].value_counts(normalize=True)

,proportion
deposit,
no,0.52616
yes,0.47384


In [22]:
#deposit rates by different fields
df.groupby("job")["deposit"].apply(lambda x: (x=="yes").mean()).sort_values(ascending=False)

,deposit
job,
student,0.747222
retired,0.663239
unemployed,0.565826
management,0.507015
Not Specified,0.485714
admin.,0.473013
self-employed,0.461728
technician,0.460779
services,0.399783


In [23]:
df.groupby("education")["deposit"].apply(lambda x: (x=="yes").mean()).sort_values(ascending=False)

,deposit
education,
tertiary,0.541068
Not Specified,0.507042
secondary,0.447407
primary,0.394000


In [24]:
df.groupby("marital")["deposit"].apply(lambda x: (x=="yes").mean()).sort_values(ascending=False)

,deposit
marital,
single,0.543491
divorced,0.481052
married,0.433790


In [26]:
# Deposit rates by balance
df.groupby("deposit")["balance"].mean()

,balance
deposit,
no,1280.227141
yes,1804.267915


In [27]:
# number of contacts vs success
df.groupby("deposit")["campaign"].mean()

,campaign
deposit,
no,2.839264
yes,2.141047


In [28]:
#distribution
df["campaign"].describe()

,campaign
count,11162.000000
mean,2.508421
std,2.722077
min,1.000000
25%,1.000000
50%,2.000000
75%,3.000000
max,63.000000


In [29]:
# duration verses subscription
df.groupby("deposit")["duration"].mean()
# higher duration most likely to subscribe

,duration
deposit,
no,223.130257
yes,537.294574


In [30]:
df["deposit"].value_counts()
df["deposit"].value_counts(normalize=True)

,proportion
deposit,
no,0.52616
yes,0.47384


**INTERPRETATION**
no   → 0.52616 (52.6%)
yes  → 0.47384 (47.4%)
The dataset is fairly balanced
Slightly more people did NOT subscribe than those who did
The deposit is relatively balanced, with 52.6% non-subscribers and 47.4% subscribers, meaning no major class imbalance adjustments were required.

In [32]:
# Saving
df.to_csv("bank_marketing_cleaned.csv", index=False)